tailscale/resource_oauth_client.go

```diff
@@ -0,0 +1,154 @@
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Optional:    true,
+				Description: "A list of tags that access tokens generated for the OAuth client will be able to assign to devices. Mandatory is the scopes include \"devices:core\" or \"auth_keys\".",
+				ForceNew:    true,
+			},
+			"id": {
+				Type:        schema.TypeString,
+				Description: "The client ID, also known as the key id. Used with the client secret to generate access tokens.",
+				Computed:    true,
+				Sensitive:   true,
+			},
+			"key": {
+				Type:        schema.TypeString,
+				Description: "The client secret, also known as the key. Used with the client ID to generate access tokens.",
+				Computed:    true,
+				Sensitive:   true,
+			},
+			"created_at": {
+				Type:        schema.TypeString,
+				Description: "The creation timestamp of the key in RFC3339 format",
+				Computed:    true,
+			},
+			"user_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "ID of the user who created this key, or empty if not created by a user.",
+				Computed:    true,
+			},
+		},
+	}
+}
+
+func resourceOAuthClientRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*tailscale.Client)
+	key, err := client.Keys().Get(ctx, d.Id())
+	if err != nil {
+		return diagnosticsError(err, "Failed to fetch key")
+	}
+
+	d.SetId(key.ID)
+	if err = d.Set("description", key.Description); err != nil {
+		return diagnosticsError(err, "Failed to set description")
+	}
+
+	if err = d.Set("created_at", key.Created.Format(time.RFC3339)); err != nil {
+		return diagnosticsError(err, "Failed to set created_at")
+	}
+
+	if err = d.Set("user_id", key.UserID); err != nil {
+		return diagnosticsError(err, "Failed to set 'user_id'")
+	}
+
+	return nil
+}
+
+func resourceOAuthClientCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	client := m.(*tailscale.Client)
+
+	description, ok := d.GetOk("description")
+	if !ok {
+		description = ""
+	}
+	var scopes []string
+	for _, scope := range d.Get("scopes").(*schema.Set).List() {
+		scopes = append(scopes, scope.(string))
+	}
+	var tags []string
+	for _, tag := range d.Get("tags").(*schema.Set).List() {
+		tags = append(tags, tag.(string))
+	}
+
+	key, err := client.Keys().CreateOAuthClient(ctx, tailscale.CreateOAuthClientRequest{
+		Description: description.(string),
+		Scopes:      scopes,
+		Tags:        tags,
+	})
+	if err != nil {
+		return diagnosticsError(err, "Failed to create oauth client")
+	}
+
+	d.SetId(key.ID)
+	if err = d.Set("key", key.Key); err != nil {
+		return diagnosticsError(err, "Failed to set key")
+	}
+	if err = d.Set("created_at", key.Created.Format(time.RFC3339)); err != nil {
+		return diagnosticsError(err, "Failed to set created_at")
+	}
+	if err = d.Set("user_id", key.UserID); err != nil {
+		return diagnosticsError(err, "Failed to set user_id")
+	}
+
+	return nil
```

Comment:
As a sanity measure we usually do a read after create to protect against potential [state drift issues](https://developer.hashicorp.com/terraform/plugin/sdkv2/best-practices/detecting-drift#update-state-after-modification):

```suggestion
	return resourceOAuthClientRead(ctx, d, m)
```